### ISSUE TITLE
Cachebust for Sass files broken in 0.14.0?

### ISSUE THREAD

**Issue by InputUsername** [getzola/zola#1570]
# Bug Report

## Environment

Zola version: 0.14.0

## Expected Behavior
Linking to a Sass stylesheet `style.scss` that is compiled to `style.css`, using `get_url(path="style.css", cachebust=true)`, should produce a URL to this file. This works in 0.12.2.

## Current Behavior
In Zola 0.14.0, the exact same `get_url(path="style.css", cachebust=true)` gives the following error:
```text
Building site...
Checking all internal links with anchors.
> Successfully checked 0 internal link(s) with anchors.
-> Creating 3 pages (0 orphan), 1 sections, and processing 0 images
Error: Failed to render page '<site path>/content/about.md'
Reason: Failed to render 'page.html' (error happened in a parent template)
Reason: Function call 'get_url' failed
Reason: `get_url`: Could not find or open file style.css
```

The error disappears when removing `cachebust=true`.

## Step to reproduce
- Create Sass stylesheet `style.scss`;
- Link to it in a template using `<link rel="stylesheet" href="{{ get_url(path='style.css', cachebust=true) }}">`

**Comment by Keats** [comment]
And it works without the cachebust? weird

**Comment by InputUsername** [comment]
I ran a `git bisect` and found this:
```
9e0cac6d4d1623918f63e08848b9a7a219621289 is the first bad commit
commit 9e0cac6d4d1623918f63e08848b9a7a219621289
Author: Vincent Prouillet <balthek@gmail.com>
Date:   Thu May 13 22:20:47 2021 +0200

    Cleanup of get_url and get_file_hash

 CHANGELOG.md                                     |   4 +-
 components/site/src/tpls.rs                      |  12 +-
 components/templates/src/global_fns/files.rs     | 435 +++++++++++++++++++++
 components/templates/src/global_fns/helpers.rs   |  36 ++
 components/templates/src/global_fns/images.rs    |   5 +-
 components/templates/src/global_fns/load_data.rs |  36 +-
 components/templates/src/global_fns/mod.rs       | 473 +----------------------
 test_site/themes/sample/templates/index.html     |   2 +-
 8 files changed, 501 insertions(+), 502 deletions(-)
 create mode 100644 components/templates/src/global_fns/files.rs
 create mode 100644 components/templates/src/global_fns/helpers.rs
```

**Comment by lcowell** [comment]
I did a little digging on this too. What I found was that if you specify an invalid filepath with `cachebust` enabled, you get the originally posted error. If cachebust is disabled, it will noop as it bypasses a bunch of additional checks `/components/templates/src/global_fns/files.rs#L113`.

I suppose it's a little confusing that if you enable `cachebust`, you get an error, but if you don't you don't (and you'll just get an error in the browser). I'd expect the behaviour to be the same, with or without the feature enabled. What do people think about ensuring we get a similar error whether `cachebust` is enabled or not? Does that seem like the right fix?

**Comment by Keats** [comment]
It's something overlooked in the refactor, it should be a noop like the rest if the file is not found.

**Comment by InputUsername** [comment]
Looks like a matter of changing [`/components/templates/src/global_fns/files.rs#L122-L128`](https://github.com/getzola/zola/blob/master/components/templates/src/global_fns/files.rs#L122-L128) to `None => {}`, correct?

Wouldn't that ignore cachebust for Sass files that haven't been compiled yet? Is that intended behavior?

**Comment by Keats** [comment]
Actually looking at the code again, we use the sha256 of the file for cachebust so it makes sense it errors if the file is not absent. Should it fallback to a timestamp in that case?

**Comment by lcowell** [comment]
My 2 cents...

I think that we should always error if there's a call to `get_url` with an unresolvable file.
* it would offer more consistent/predictable behaviour
* it's doesn't increase the complexity of the system
* could potentially save people from publishing sites with broken links (they'll still be broken in the browser).
* reduce bug reports (I bet @InputUsername wouldn't have filed this as a bug if the behaviour had been the same with `cachebust` on or off).

I haven't been using zola for that long, so happy to defer to others with more experience here.

**Comment by Keats** [comment]
> I think that we should always error if there's a call to get_url with an unresolvable file

No, because some people generate some JS files through some build system for example and the JS step might be ran before or after zola and it should work in both cases hopefully.

I believe the easy fix here is to move the Sass generation at the started of the build but cachebust will still be an issue for those assets coming from other systems.

**Comment by InputUsername** [comment]
My personal preference is to have cachebust working for Sass stylesheets :sweat_smile: 

~~Other than that, I agree that consistent error behavior for nonexistent files sounds like a good idea for preventing broken links.~~

Edit: just saw Keats's reply

**Comment by lcowell** [comment]
> No, because some people generate some JS files through some build system for example and the JS step might be ran before or after zola and it should work in both cases hopefully.

This is a great point and I wouldn't want to break any supported use cases.

I was thinking this through and realized that even if we allowed the JS build step to happen after the `zola build` step, zola would be fingerprinting the previous version of the file when it generates the html. I don't think is desirable either as it could result in an updated js file not getting its cache busted.

I wonder how easy it would be for users of the use-case you'd mentioned to orchestrate things in a logical order? Does anyone know how this is handled in similar projects?

**Comment by Keats** [comment]
Good point, not entirely sure what to do there

**Comment by chris-morgan** [comment]
I just hit this in rebasing my personal fork from 0.10.2+dev, a task which I’m now going to shelve until this is resolved.

In reading the docs, I saw mentions of get_url with path beginning with “static/”, so I figured maybe I need to do sass/foo.scss, and it’ll perform the mapping to /foo.css itself—which wouldn’t seem odd as it’s basically what it does with paths like @/foo.md. But it turned out that it was calculating the hash of the source file sass/foo.scss, and then producing a link to base_url/sass/foo.scss, the combination of which is obviously Wrong.

get_url needs to decide where it’s operating, input or output, and be consistent about it. I see cache busting as unequivocally about output paths; any other build stuff must *obviously* happen before Zola if cache busting is employed, writing its built artefacts to static/, sass/, content/ or wherever.

I’m not thinking too deeply about it now, but it’s *possible* that get_url should even be split into two functions, one operating on inputs and one on outputs—one taking static/foo.jpg and maybe even sass/foo.scss, and another taking foo.jpg and foo.css.

**Comment by lcowell** [comment]
It kind of seems like `get_url` is doing 2 things:
* resolving the file path
* generating a url

I think it would be interesting to break it out in to 2 separate operations (at least internally). I was wondering if, given that this thread is already getting complicated, whether that should be broken out in to a separate issue? that way this doesn't get lost in this discussion.

In order to support the workflow where people want to generate the dependent files (like JS) after zola builds, I think we could add an additional flag to `get_url` like `ignore_errors` or something like that (name is just an example). This flag should be incompatible with the `cachebust` flag as you can't generate a hash on a file that doesn't exist yet.

**Comment by Keats** [comment]
Actually it's just that I completely forgot to add the `public` dir ot the list of folders to check in for all the files related Tera function >_>
So that's an easy fix at least, I might get to it this weekend but essentially the output path needs to be added in https://github.com/getzola/zola/blob/master/components/templates/src/global_fns/helpers.rs#L7-L21
It also needs to be passed around from the `Site` since it's configurable.

**Comment by InputUsername** [comment]
@Keats I have some time to work on this, do you mind if I make a PR?

**Comment by Keats** [comment]
Go for it

**Comment by biodranik** [comment]
get_url(path="sass/main.scss") produces invalid link to baseurl/sass/main.scss even without cache boost. This is a broken behavior, right?

**Comment by Keats** [comment]
By default it will still generate the link even if it doesn't find a file, that's the expected behaviour since it's not sure zola will handle or be aware of all the content/pages/files.
However if the cachebust option is on, it should error since we can't do it without the file.

**Comment by Keats** [comment]
Should be fixed in 0.14.1
